### PR TITLE
Fixed GraphQL Get

### LINKED ISSUE #288: GraphQL Get: Query all properties if none are given
Currently get queries require to list all properties that should be returned and if no properties are given, the query returns an error. 

```
client.query.get(class_name="CLASS", properties=["PROP1", "PROP2"]).do()
```

Instead of returning an error, the client should get all non-reference properties from the schema if `properties=None` and include those by default

### PR DESCRIPTION
**Pull request by meetagrawal09**
This PR fixes #288

### CODE REVIEW

**Comment by dirkkul** [comment]
thank you for your PR! 

Could you add a test? Best would be an integration test that includes references

File: integration/test_graphql.py [hunk 1]
```diff
@@ -89,6 +89,15 @@ def test_get_data(client):
...
 
+def test_get_data_with_properties_default(client):
+    """Test GraphQL's Get clause with properties default (None)."""
+    result = client.query.get("Ship").with_limit(2).do()
+    objects = get_objects_from_result(result)
+    all_props_found = True
+    for obj in objects:
+        if not obj["name"] or not obj["size"] or not obj["description"]:
+            all_props_found = False
+    assert all_props_found and len(objects) == 2
```
Comment (dirkkul) [inline integration/test_graphql.py:100]: thanks for the test, but there are no references in the ship-schema. You could reuse the people schema?

File: weaviate/gql/get.py [hunk 2]
```diff
@@ -83,7 +83,13 @@ def __init__(
         if not isinstance(class_name, str):
             raise TypeError(f"class name must be of type str but was {type(class_name)}")
         if properties is None:
-            properties = []
+            class_schema = client.Schema.get(class_name=class_name)
```
Comment (dirkkul) [inline weaviate/gql/get.py:86]: you don'T have access to the client here.

What you could do:
- When instantiating query, you could pass the schema object as additional argument
- Then you should be able to do schema.get(class_name)

File: integration/test_graphql.py [hunk 3]
```diff
@@ -89,6 +89,34 @@ def test_get_data(client):
...
+        },
+        "Person",
+    )
+
+    client.data_object.create(
+        {
+            "name": "Group-Weaviate",
+            "members": [
+                {
+                    "beacon": "weaviate://localhost/" + str(person_id),
```
Comment (meetagrawal09) [inline integration/test_graphql.py:109]: @dirkkul, I need some help regarding how can I create this reference correctly.

File: integration/test_graphql.py (hunk 1 above)
Comment (meetagrawal09) [reply integration/test_graphql.py]: I have moved to using `Group` class which has a reference property to `Person` in the `people_schema`.

File: integration/test_graphql.py (hunk 3 above)
Comment (dirkkul) [reply integration/test_graphql.py:109]: First add the group without references (eg leave the members out) and then use data_object.add_reference() toa dd the reference. You can check integration/test_crud.py for examples

Comment (meetagrawal09) [reply integration/test_graphql.py]: Thanks, made the updates.

File: weaviate/gql/get.py [hunk 4]
```diff
@@ -83,7 +84,13 @@ def __init__(
         if not isinstance(class_name, str):
             raise TypeError(f"class name must be of type str but was {type(class_name)}")
         if properties is None:
-            properties = []
+            class_schema = schema.get(class_name=class_name)
+            if class_schema:
```
Comment (dirkkul) [inline weaviate/gql/get.py:88]: What exactly is the ìf` for?
Afaik get returns an exception if the class does not exists

File: weaviate/gql/get.py [hunk 5]
```diff
@@ -83,7 +84,13 @@ def __init__(
...
+                non_reference_properties = [
+                    prop for prop in class_schema["properties"] if prop["dataType"][0][0].islower()
+                ]
+                properties = [prop["name"] for prop in non_reference_properties]
```
Comment (dirkkul) [inline weaviate/gql/get.py:92]: I think we can directly do 

```
                non_reference_properties = [
                    prop["name"] for prop in class_schema["properties"] if prop["dataType"][0][0].islower()
                ]
```

File: weaviate/gql/get.py (hunk 4 above)
Comment (meetagrawal09) [reply weaviate/gql/get.py]: Removed it.

File: weaviate/gql/get.py (hunk 5 above)
Comment (meetagrawal09) [reply weaviate/gql/get.py]: Yes, made the changes.

**Comment by meetagrawal09** [comment]
> To avoid any confusion in the future about your contribution to Weaviate, we work with a [Contributor License Agreement](https://weaviate.io/service/contributor-license-agreement). If you agree, you can simply add a comment to this PR that you agree with the CLA so that we can merge.beep boop - the Weaviate bot 👋🤖PS:Are you already a member of the [Weaviate Slack channel?](https://weaviate.io/slack)

I agree with the Contributor License Agreement.

**Comment by meetagrawal09** [comment]
What should be the behavior in case a `class_name` does not have any non-reference dataType properties ?

**Comment by dirkkul** [comment]
> What should be the behavior in case a class_name does not have any non-reference dataType properties ?

I'm not 100% sure if I understood that. Do you mean if the class returned by the schema call does not contain a datatype?

**Comment by meetagrawal09** [comment]
@dirkkul, if the below schema is used : 

 ```
[
   {
      "class": "Person",
      "description": "A person such as humans or personality known through culture",
      "properties": [
        {
          "name": "name",
          "description": "The name of this person",
          "dataType": [
            "string"
          ]
        }
      ]
    },
    {
      "class": "Group",
      "description": "A set of persons who are associated with each other over some common properties",
      "properties": [ 
        {
          "name": "members",
          "description": "The persons that are part of this group",
          "dataType": [
            "Person"
          ]
        }
      ]
    },
  ]
```

and then we do `client.query.get("Group")`, what should be a returned value ? (If you see group has got just one property that is a `reference` property and no `non-reference` property)

**Comment by dirkkul** [comment]
ahh, that is a good question. Currently it is run without any properties and will fail, right?

File: weaviate/gql/get.py [hunk 6]
```diff
@@ -83,7 +84,12 @@ def __init__(
         if not isinstance(class_name, str):
             raise TypeError(f"class name must be of type str but was {type(class_name)}")
         if properties is None:
-            properties = []
```
Comment (dirkkul) [inline weaviate/gql/get.py:86]: Can you adept the docstring?

**Comment by meetagrawal09** [comment]
> ahh, that is a good question. Currently it is run without any properties and will fail, right?

Yes exactly

**Comment by dirkkul** [comment]
I think this is ok.

It failed before, too. So I think it still is an improvement over the previous state

File: weaviate/gql/get.py (hunk 6 above)
Comment (meetagrawal09) [reply weaviate/gql/get.py:86]: Done.

Comment (dirkkul) [reply weaviate/gql/get.py:86]: that is not the docstring

Needs to be added here:
```
        Parameters
        ----------
        class_name : str
            Class name of the objects to interact with.
        properties : str or list of str
            Properties of the objects to interact with. <=== here
```

Comment (meetagrawal09) [reply weaviate/gql/get.py:86]: Added the `schema` parameter to it at both the required places.

**Comment by meetagrawal09** [comment]
@dirkkul Since the PR has added a new parameter `schema` to be passed to `GetBuilder`, we actually need to update the older calls made to the constructor of this class i.e. tests for `GetBuilder` need to be updated. One such occurence is here : [Link](https://github.com/weaviate/weaviate-python-client/blob/main/test/gql/test_get.py#L100)

File: weaviate/gql/get.py (hunk 6 above)
Comment (dirkkul) [reply weaviate/gql/get.py:86]: I meant the properties entry, that needs an addition that `None` means "all properties"

Comment (dirkkul) [reply weaviate/gql/get.py:86]: Also in query.get()

**Comment by dirkkul** [comment]
Hi, you're right about adding it in the unit-tests. I think just adding a mock there would be fine

**Comment by dirkkul** [comment]
I noticed one more thing:

we only want to return all properties if there are no additional properties.

```
query.get("Class", None) => returns all properties
query.get("Class", None).with_additional("id") => returns only _additional(id) and no other properties
```

I think the correct place for this would be to move the "empty check" to here https://github.com/weaviate/weaviate-python-client/blob/1f1f80984a43ca9de2bf58ff9fe3b4a5daa8fb94/weaviate/gql/get.py#L1055

**Comment by meetagrawal09** [comment]
@dirkkul, I have made all the corrections since the last review. You can please proceed with adding comments about any further improvements that might be needed.

File: integration/test_graphql.py [hunk 7]
```diff
@@ -108,6 +108,44 @@ def test_get_data(client):
...
+        to_class_name="Person",
+    )
+
+    result_without_additional_props = client.query.get("Group").with_limit(1).do()
+    result_with_additional_props = client.query.get("Group").with_limit(1).with_additional(["id"]).do()
+    objects_without_additional_props = get_objects_from_result(result_without_additional_props)
+    objects_with_additional_props = get_objects_from_result(result_with_additional_props)
+
+    assert "name" in objects_without_additional_props[0]
+    assert "id" in objects_with_additional_props[0]["_additional"]
```
Comment (dirkkul) [inline integration/test_graphql.py:147]: This check also needs to include that "name" is NOT part of the answer

Comment (meetagrawal09) [reply integration/test_graphql.py:147]: Added

**Comment by dirkkul** [comment]
closing this as the tests are still failing and there hasn't been any activity in almost a month

**Pull request closed without merging**